keystrokes/src/components/Character/CharacterDetails/ManualKeybindInputs.js

```diff
@@ -0,0 +1,368 @@
+import React from 'react';
+import { red } from '@material-ui/core/colors';
+import {
+	Tooltip,
+	Button,
+	Grid,
+	TextField,
+	Typography,
+	MenuItem,
+	Avatar,
+} from '@material-ui/core';
+import WarningIcon from '@material-ui/icons/Warning';
+import PropTypes from 'prop-types';
+import styleGuide from '../../../stylesheets/style';
+import { verifyKey, validatePress } from '../../utils/utils';
+import { targetting, mods } from '../../../config/constants';
+
+ManualKeybindInputs.propTypes = {
+	spec: PropTypes.string.isRequired,
+	invalidBind: PropTypes.array.isRequired,
+	setInvalidBind: PropTypes.func.isRequired,
+	onSubmit: PropTypes.func.isRequired,
+	allKeybinds: PropTypes.array.isRequired,
+	keybinding: PropTypes.object.isRequired,
+	setKeybinding: PropTypes.func.isRequired,
+	checkIfInvalidAndAdd: PropTypes.func.isRequired,
+	Spells: PropTypes.object.isRequired,
+};
+
+function ManualKeybindInputs({
+	spec,
+	invalidBind,
+	setInvalidBind,
+	onSubmit,
+	allKeybinds,
+	keybinding,
+	setKeybinding,
+	checkIfInvalidAndAdd,
+	Spells,
+}) {
+	const classes = styleGuide();
+
+	function handleKeyPress(e) {
+		if (!e.metaKey) {
+			e.preventDefault();
+		}
+		const newKey = e.code.toLowerCase().replace(/digit|key/i, '');
+		if (validatePress(newKey)) {
+			setKeybinding({
```

Comment:
I'd personally put the condition on the one key property since that the variable thing. Could remove like 4-6 lines

Reply:
interesting point, i made a good bunch of changes here to handle this, now its substantially shorter.

Reply:
getting some weird behavior when i change this around. I changed it to be 
function handleKeyPress(e) {
		if (!e.metaKey) {
			e.preventDefault();
		}
		const newKey = e.code.toLowerCase().replace(/digit|key/i, '');
		const newKeybinding = {
			...keybinding,
			Key: validatePress(newKey) ? verifyKey(newKey) : keybinding.Key,
		};
		setKeybinding(newKeybinding);
		checkIfInvalidAndAdd(newKeybinding);
	}

But for some reason the key doesnt seem to be set until you click off to remove the function from onkeydown. The weird thing is it still shows the correct key update, but it just doesnt seems to capture it in the function. I think it might have to do with 

				<TextField
					className={classes.button}
					disabled={!keybinding.Spell}
					variant="outlined"
					id="keystroke-selector"
					value={keybinding.Key || ''}
					label="Key"
					onFocus={() => {
						document.body.onkeydown = handleKeyPress;
					}}
					onBlur={() => {
						document.body.onkeydown = null;
					}}
				/>
Since i think setting functions this way does something weird with state. Either way, im not sure exactly what to do to fix it, but i can just make a few changes to make the behaviour remain unchanged. Lemme know if u have any insight into why this is happening, would be great to know